### PR TITLE
fix: make getBoundTextElement and related helpers pure

### PR DESCRIPTION
**Pull request by ad1992**
- [x] Make `getBoundTextElement` pure by removing `scene ` and passing `elementsMap` instead
- [x] Pass `elementsMap` to `computeBoundTextPosition` `getContainerCenter` and `redrawTextBoundingBox`
- [x] Pass `elementsMap` to `hitTest`, `handleBindTextResize`, `ElementBounds.calculateBounds` , `rotateMultipleElements`
- [x] Pass `elementsMap` to `getLinearElementRotatedBounds` and `LinearElementEditor.getEditorMidPoints`
- [x] Pass `boundTextElement` to `getBoundTextMaxWidth`
- [x] Pass `elementsMap` to `removeElementsFromFrame`
- [x] Pass `elementsMap` to `alignElements`, `distributeElements` and `getMaximumGroups`
- [x]  Pass allElementsMap to renderElement and StaticCanvas

### CODE REVIEW

File: packages/excalidraw/element/textElement.ts [hunk 1]
```diff
@@ -877,10 +894,12 @@ export const computeContainerDimensionForBoundText = (
 
 export const getBoundTextMaxWidth = (
   container: ExcalidrawElement,
-  boundTextElement: ExcalidrawTextElement | null = getBoundTextElement(
-    container,
-  ),
+  boundTextElement: ExcalidrawTextElement | null,
+  elementsMap: ElementsMap = new Map(),
```
[the remark this answers is not in the document]
Comment (ad1992) [reply packages/excalidraw/element/textElement.ts:898]: > why pass elementsMap when we're already passing boundTextElement?
 Good point, in that case I think I can just create a map and pass it to `getBoundTextElement` which is being called from the above function

**Comment by ad1992** [comment]
@dwelle do you know what was the reason for passing `elementsMap` in [transformElements](https://github.com/excalidraw/excalidraw/blob/678bb2b8192975c935cf313f9cf9a7837dd03d37/packages/excalidraw/element/resizeElements.ts#L75) ? We can use `arrayToMap(originalElements)` instead

**Comment by dwelle** [comment]
> @dwelle do you know what was the reason for passing `elementsMap` in [transformElements](https://github.com/excalidraw/excalidraw/blob/678bb2b8192975c935cf313f9cf9a7837dd03d37/packages/excalidraw/element/resizeElements.ts#L75) ? We can use `arrayToMap(originalElements)` instead

mainly perf (we don't want to be recreating a map on every mousemove if we don't have to) and quite possibly similar concerns to notes above, where may need to have access to all elements, not just what's currently selected.

File: packages/excalidraw/element/textWysiwyg.tsx [hunk 2]
```diff
@@ -366,14 +364,14 @@ export const textWysiwyg = ({
         fontFamily: app.state.currentItemFontFamily,
       });
       if (container) {
+        const boundTextElement = getBoundTextElement(
+          container,
+          app.scene.getNonDeletedElementsMap(),
+        )!;
```
Comment (dwelle) [inline packages/excalidraw/element/textWysiwyg.tsx:370]: doesn't need assertion here

File: packages/excalidraw/renderer/renderElement.ts [hunk 3]
```diff
@@ -457,6 +458,9 @@ const drawElementFromCanvas = (
   const zoom = elementWithCanvas.scale;
   let [x1, y1, x2, y2] = getElementAbsoluteCoords(element);
 
+  // need to use scene here since elementsMap is outdated when passed via args and breaks labelled arrows
+  const elementsMap = Scene.getScene(element)!.getNonDeletedElementsMap();
```
Comment (dwelle) [inline packages/excalidraw/renderer/renderElement.ts:462]: `elementsMap` isn't outdated. It's just a subset that doesn't contain `appState.editingElement`.

File: packages/excalidraw/actions/actionFlip.ts [hunk 4]
```diff
@@ -67,7 +68,7 @@ export const actionFlipVertical = register({
 
 const flipSelectedElements = (
   elements: readonly ExcalidrawElement[],
-  elementsMap: NonDeletedElementsMap,
+  elementsMap: NonDeletedElementsMap | NonDeletedSceneElementsMap,
```
Comment (dwelle) [inline packages/excalidraw/actions/actionFlip.ts:71]: annoying but necessary for now. We can later remove this by removing the brand on `NonDeletedElementsMap` once we make `NonDeleted<>` type into a branded type of its own.

**Review by dwelle — APPROVED** [review]

**Pull request merged**